### PR TITLE
fix: check mempool when call `eth_getTransactionByHash`

### LINKED ISSUE #1523: Send raw transaction successfully but failed to get it by its hash through web JSON-RPC apis.
### Description

When developers develops a Dapp, after sending a transaction, they may want to get the status of that transaction.

But it will be failed.

#### Why it doesn't occur in CI tests?

- The abnormal results are ignored. CI doesn't check them.

  ```javascript
  for (let i = 0; i < count; i++) {
      response = await provider.getTransactionReceipt(txHash);
      if (response == null) {
          await sleep(2000)
          continue;
      }
  ```

  Ref: [`getTxReceipt(..)` in axonweb3/axon-test](https://github.com/axonweb3/axon-test/blob/0e2e379ac17175c301115127ea679f1d0085ddc3/test/utils/rpc.js#L125-L139)

- [Hardhat] is a JavaScript toolchain, it is very slow.

  - I tried the simplest example with [Hardhat] and Axon, it's almost failed 5%.
  
  - I tried the simplest example with [Foundry] and Axon, it's almost failed 99.9%.

  - I tried the simplest example with [Foundry] and [Geth], it's 100% success.

  Due to this issue, use [Foundry] to test Axon is impossible at present.

  [Foundry] allows pending status and it will retry for pending transactions but it doesn't allow null status, it will return the error "Transaction dropped from the mempool".

  I don't mean developers have to use [Foundry], but it means some high performance tools will not support Axon.

[Hardhat]: https://hardhat.org/
[Foundry]: https://getfoundry.sh/
[Geth]: https://geth.ethereum.org/

### Reason

- After call `eth_sendRawTransaction`, the RPC will return a response after inserting that transaction into mempool.

  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/api/src/jsonrpc/impl/web3.rs#L313-L318
  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/api/src/adapter.rs#L70-L76
  
  <details><summary>Click here for more details.</summary><br/>
  
  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/mempool/src/lib.rs#L198
  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/mempool/src/lib.rs#L118-L122
  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/mempool/src/pool.rs#L135
  
  </details>

- But when call `eth_getTransactionByHash`, `eth_getTransactionReceipt` or some other APIs, the RPC will read the status from storage directly.

  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/api/src/jsonrpc/impl/web3.rs#L323-L327
  https://github.com/axonweb3/axon/blob/ad4375b415d28b4286ac51874102ac74c99c8cb2/core/api/src/adapter.rs#L139-L145

- (I tested!) 3~5 seconds after sending the transaction successfully, there will still be a chance that the transaction is still in the mempool and hasn't been inserted into storage.

### Proposal Solution

- Shouldn't return the transaction state from storage directly, check the mempool at first.
- Better tests.

### PR DESCRIPTION
**Pull request by chaoticlonghair**
## What this PR does / why we need it?

This PR resolves #1523 partially.

This PR requires #1530.

_p.s. I say "partially" because no tests are updated or added._

### What is the impact of this PR?

No Breaking Change

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [ ] Web3 Compatible Tests
- [x] OCT 1-5 And 12-15
- [x] OCT 6-10
- [x] OCT 11
- [x] OCT 16-19
- [ ] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

**Review by driftluo — DISMISSED** [review]

**Review by Flouse — DISMISSED** [review]
> _p.s. I say "partially" because no tests are updated or added._

You said that you used hardhat and Foundry to test https://github.com/axonweb3/axon/issues/1523.
So, this PR is manually tested, right?

**Comment by chaoticlonghair** [comment]
> So, this PR is manually tested, right?

Yep!

**Review by KaoImin — DISMISSED** [review]

**Comment by chaoticlonghair** [comment]
The tests should be fixed before this PR merged. (ref: https://github.com/axonweb3/axon/issues/1523#issuecomment-1794356681)

**Review by sunchengzhu — DISMISSED** [review]

**Comment by chaoticlonghair** [comment]
Hold! Beacuse the fix for v3-core tests is stilling on processing.

**Comment by Flouse** [comment]
> v3-core tests

- link https://github.com/axonweb3/axon/pull/1541

**Comment by chaoticlonghair** [comment]
Could NOT be passed since #1544.

Without the bugfix of #1544, all tests should be patched: waiting for all previous transactions mined before sending a new transaction.

**Comment by chaoticlonghair** [comment]
@Flouse Do not do squash merging.
Keep the two commits to be separated.
The 2nd commit is a temporary solution and it could be reverted after the #1544 fixed.

File: .github/workflows/web3_compatible.yml [hunk 1]
```diff
@@ -105,7 +105,7 @@ jobs:
         uses: actions/checkout@v4
         with:
           repository: axonweb3/axon-test
-          ref: 0e2e379ac17175c301115127ea679f1d0085ddc3
+          ref: be8b502423fa41d4656ec2369d3ee5a729f56326
```
Comment (Flouse) [inline .github/workflows/web3_compatible.yml:108]: https://github.com/axonweb3/axon-test/commits/be8b502423fa41d4656ec2369d3ee5a729f56326

**Review by Flouse — APPROVED** [review]

**Pull request merged**